internal/provider/data_source_gitlab_project_membership_test.go

```diff
@@ -0,0 +1,98 @@
+package provider
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/acctest"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+)
+
+func TestAccDataSourceGitlabProjectMembership_basic(t *testing.T) {
+	rInt := acctest.RandInt()
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:          func() { testAccPreCheck(t) },
+		ProviderFactories: providerFactories,
+		Steps: []resource.TestStep{
+			// Create the project and one member
+			{
+				Config: testAccDataSourceGitlabProjectMembershipConfig(rInt),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("gitlab_project.foo", "name", fmt.Sprintf("foo%d", rInt)),
+					resource.TestCheckResourceAttr("gitlab_user.test", "name", fmt.Sprintf("foo%d", rInt)),
+					resource.TestCheckResourceAttr("gitlab_project_membership.foo", "access_level", "developer"),
+				),
+			},
+			{
+				Config: testAccDataSourceGitlabProjectMembershipConfig_basic(rInt),
+				Check: resource.ComposeTestCheckFunc(
+					// Members is 2 because the user owning the token is always added to the project
+					resource.TestCheckResourceAttr("data.gitlab_project_membership.foo", "members.#", "2"),
+					resource.TestCheckResourceAttr("data.gitlab_project_membership.foo", "members.1.username", fmt.Sprintf("listest%d", rInt)),
+				),
+			},
+		},
+	})
+}
+
+func TestAccDataSourceGitlabProjectMembership_pagination(t *testing.T) {
+	testAccCheck(t)
+
+	userCount := 21
+
+	project := testAccCreateProject(t)
+	users := testAccCreateUsers(t, userCount)
+	testAccAddProjectMembers(t, project.ID, users)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:          func() { testAccPreCheck(t) },
+		ProviderFactories: providerFactories,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccDataSourceGitlabProjectMembershipPagination(project.ID),
+				Check:  resource.TestCheckResourceAttr("data.gitlab_project_membership.this", "members.#", fmt.Sprintf("%d", userCount)),
+			},
+		},
+	})
+}
+
+func testAccDataSourceGitlabProjectMembershipConfig(rInt int) string {
+	return fmt.Sprintf(`
+resource "gitlab_project" "foo" {
```

Comment:
Let's only have what's "under test" in the test config and use the `helper_test.go` functions to create the necessary test setup.

Reply:
Just to make sure I understand this correctly, you want me to move the function `testAccDataSourceGitlabProjectMembershipConfig` (that's a mouthful!) to the `helpers_test.go` file?

Reply:
Not exactly, but e.g. use the `testAccCreateProject()` function from that module to setup the project before the test steps. Something along the lines of this: 

https://github.com/gitlabhq/terraform-provider-gitlab/blob/d18aee439f62646265b5dd0b954f500449cc85a8/internal/provider/resource_gitlab_project_issue_test.go#L20